pandas/tests/indexes/multi/test_setops.py

```diff
@@ -754,13 +752,12 @@ def test_intersection_keep_ea_dtypes(val, any_numeric_ea_dtype):
     tm.assert_index_equal(result, expected)
 
 
-@pytest.mark.xfail(using_string_dtype(), reason="TODO(infer_string)")
 def test_union_with_na_when_constructing_dataframe():
     # GH43222
     series1 = Series(
         (1,),
         index=MultiIndex.from_arrays(
-            [Series([None], dtype="string"), Series([None], dtype="string")]
+            [Series([None], dtype="str"), Series([None], dtype="str")]
```

Comment:
So this does fix the test. Is the behavior if this change to the test is not made correct?

i.e. the `series1.index.dtypes` are `object` and the `series2.index.dtypes` are `str` and the resulting dtype for the columns index using the DataFrame constructor is `object`. Would we not expect the DataFrame constructor to return a `str` index for the columns in this case?

Reply:
> i.e. the `series1.index.dtypes` are `object` and the `series2.index.dtypes` are `str` and the resulting dtype for the columns index using the DataFrame constructor is `object`.

With the above fix (and when infer_string is enabled), the test uses `str` dtype for the index levels of both series1 and series2, and then also the expected result gets created with that. 

So it's only testing that the NaNs are properly matched when creating the rows from Series objects with a MultiIndex, it does not test having different dtypes in series1 vs series2.

That's also something we could test, though (and then if you have object dtype index and str dtype index, one would expect the result to be object dtype index (since that is the "common" dtype), but was not how the test was currently set up.

> perhaps related to https://github.com/pandas-dev/pandas/issues/60338?

I suppose not because here we don't actually have empty objects. Both `series1` and `series2` have data and have a non-empty index.